_security/access-control/resources.md

```diff
@@ -0,0 +1,173 @@
+```
+{% include copy.html %}
+
+For information about implementing this feature in your plugin, see [Resource access control for plugins](https://github.com/opensearch-project/security/blob/main/RESOURCE_ACCESS_CONTROL_FOR_PLUGINS.md).
+{: .tip }
+
+### Cluster permissions
+
+The Security plugin must have the following access:
+
+- Permissions to make share, verify, and list requests.
+- Shared access to all cluster components, or be the owner of the cluster.
+
+To grant the resource sharing extension these permissions, add the following role to `roles.yaml`:
+
+```yaml
+sample_full_access:
+  cluster_permissions:
+    - 'cluster:admin/sample-resource-plugin/*'
+
+sample_read_access:
+  cluster_permissions:
+    - 'cluster:admin/sample-resource-plugin/get'
+```
+{% include copy.html %}
+
+## Resource sharing components
+
+The resource sharing extension consists of key components that work together to provide standardized access management. The primary component is the Service Provider Interface (SPI), which serves as the foundation for plugin integration and resource management.
+
+### opensearch-security-spi
+
+The `opensearch-security-spi` component:
+
+- Defines `ResourceSharingExtension` for plugin implementations.
+- Tracks registered resource plugins at startup.
+- Exposes a `ResourceSharingClient` for performing share, revoke, verify, and list operations.
+
+You can customize this component based on the SPI implementation. To customize, create an extension file `src/main/resources/META-INF/services/org.opensearch.security.spi.ResourceSharingExtension` containing your SPI's class name:
+
+```
+com.example.MyResourceSharingExtension
+```
+{% include copy.html %}
+
+### API
+
+All sharing metadata is stored in the `.opensearch_resource_sharing` system index, as shown in the following table.
+
+| Field | Type | Description |
+| :--- | :--- | :--- |
+| `source_idx`  | String | The system index holding the resource |
+| `resource_id` | String | The resource ID   |
+| `created_by`  | Object | The name of the user who created the resource    |
+| `share_with`  | Object | A map of `action-groups` to access definitions  |
+
+#### Java client APIs
+
+Use the following classes when interacting with the extension through the Java client:
+
+```java
+ResourceSharingClient client = ResourceSharingClientAccessor.getResourceSharingClient();
+
+// Verify access
+client.verifyResourceAccess(resourceId, indexName, listener);
+
+// Share resource
+client.share(resourceId, indexName, recipients, listener);
+
+// Revoke access
+client.revoke(resourceId, indexName, recipients, listener);
+
+// List accessible IDs
+client.getAccessibleResourceIds(indexName, listener);
+```
+{% include copy.html %}
+
+### Document structure
+
+The resource sharing metadata is stored as JSON documents that define ownership, permissions, and access patterns. Each document contains fields that specify the resource location, identifier, creator information, and sharing configuration:
```

Comment:
Where would I find or add this document? Is it an API call to the security plugin?

Reply:
This is not visible to regular users directly as it is hidden behind SystemIndex Protection.

For plugins intending to declare a protected resource, they can retrieve the information about who can access this resources by simply calling security client's hasPermission.

End users will, in future, iterations be able to control who they can share the resource with once plugin implements those APIs and makes the info available via dashboard.

At this point, the documentation intends to serve plugin developers, with allowing admins to optionally opt in to this feature by enabling feature flag.

We are trying to implement a sharing model that doesn't exist in OpenSearch yet. This is the first step by centralizing the sharing information to security, rather than plugins storing it individually.

Reply:
What would the `hasPermission` call look like?

Reply:
Plugins would call the client by passing the id of the resources, the index name where the resource sits and a listener to receive the response. This will then be evaluated on security plugin side and a boolean response will be sent through the listener.